types/src/front/acl.ts

```diff
@@ -1,29 +1,84 @@
 import { ModelId } from "../shared/model_id";
+import { GroupType } from "./groups";
+import { LightWorkspaceType, RoleType } from "./user";
 
 // Supported permissions
-export const SUPPORTED_PERMISSIONS = ["read", "write"] as const;
+export const SUPPORTED_PERMISSIONS = ["admin", "read", "write"] as const;
 
 export type Permission = (typeof SUPPORTED_PERMISSIONS)[number];
 
-// Access Control Entry
-export type ACEType = {
-  groupId: ModelId;
+export type GroupPermission = {
+  id: ModelId;
   permissions: Permission[];
 };
 
-// Access Control List
-export type ACLType = {
-  aclEntries: Array<ACEType>;
+export type RolePermission = {
+  name: RoleType;
+  permissions: Permission[];
+};
+
+export type GroupOnlyACL = {
+  groups: GroupPermission[];
+  roles?: never[];
 };
 
-export function groupHasPermission(
+export type RoleBasedACL = {
+  groups: GroupPermission[];
+  roles: RolePermission[];
+  workspaceId: ModelId; // Required when roles are defined.
+};
+
+function hasRoles(acl: ACLType): acl is RoleBasedACL {
+  return "roles" in acl;
+}
+
+export type ACLType = GroupOnlyACL | RoleBasedACL;
+
+/**
+ * Determines if a user has a specific permission on a resource based on their role and group memberships.
+ *
+ * Permission is granted if either:
+ * 1. Role-based access:
+ *    - Resource has role-based ACLs
+ *    - Either:
+ *      • Resource has public access for the permission
+ *      • User's role has the permission AND resource belongs to user's workspace
+ * 2. Group-based access:
+ *    - User belongs to a group that has the permission
+ */
+export function hasResourcePermission(
   acl: ACLType,
+  workspace: LightWorkspaceType,
   permission: Permission,
-  groupId: ModelId
+  groups: GroupType[],
+  userRole: RoleType
 ): boolean {
-  const entry = acl.aclEntries.find((ace) => ace.groupId === groupId);
-  if (entry) {
-    return entry.permissions.includes(permission);
+  // Check role-based permissions if applicable.
+  if (hasRoles(acl)) {
+    // Public access check (across all workspaces).
+    const publicPermission = acl.roles
+      .find((r) => r.name === "none")
+      ?.permissions.includes(permission);
+    if (publicPermission) {
+      return true;
+    }
+
+    // Workspace-specific role permission check.
+    const hasRolePermission = acl.roles.some(
+      (r) => userRole === r.name && r.permissions.includes(permission)
+    );
+
+    if (hasRolePermission && workspace.id === acl.workspaceId) {
+      return true;
+    }
   }
-  return false;
+
+  // Group-based permission check.
+  return groups.some((userGroup) =>
+    acl.groups.some(
+      (aclGroup) =>
+        aclGroup.id === userGroup.id &&
+        aclGroup.permissions.includes(permission)
+    )
+  );
```

Comment:
For simplicity, I would compute the union of all permissions that I have and then check if the requested permission in in it.